src/waitress/channel.py

```diff
@@ -153,6 +161,9 @@ def handle_read(self):
         if data:
             self.last_activity = time.time()
             self.received(data)
+        else:
+            # Client disconnected.
+            self.connected = False
```

Comment:
I'm struggling to understand why this is being added / why it's necessary now versus before.

Reply:
The idea of this PR is to detect a client disconnect even before the application starts to write the response, so the application may check at some strategic points if the client is still interested in the answer.

In a `select` loop, a socket that has been closed by the other side makes itself known by being readable (it is included in the readable list of the result of `select`), but if we read something, we get an empty answer. Therefore, we keep the channel `readable` (at least if there are not too many requests already queued from the channel) and if `handle_read` is then called and the read yields an empty response, this means the client has disconnected.